src/aspire/orientation/__init__.py

```diff
@@ -1,38 +1,556 @@
-        constructor of an object for estimating 3D orientations
+        Estimate 2D shifts in images using estimated rotations
+
+        The function computes the common lines from the estimated rotations,
+        and then, for each common line, estimates the 1D shift between its two
+        Fourier rays (one in image i and one in image j). Using the common
+        lines and the 1D shifts, the function solves the least-squares
+        equations for the 2D shifts.
+        This function processes the (Fourier transformed) images exactly as the
+        `build_clmatrix` function.
+
+        :param memory_factor: If there are N projections, then the system of
+            equations solved for the shifts is of size 2N x N(N-1)/2 (2N
+            unknowns and N(N-1)/2 equations). This may be too big if N is
+            large. If `memory_factor` between 0 and 1, then it is the
+            fraction of equation to retain. That is, the system of
+            equations solved will be of size 2N x N*(N-1)/2*`memory_factor`.
+            If `memory_factor` is larger than 100, then the number of
+            equations is estimated in such a way that the memory used by the equations
+            is roughly `memory_factor megabytes`. Default is 10000 (use all equations).
+            The code will generate an error if `memory_factor` is between 1 and 100.
+        :return: Estimated shifts for all images
         """
-        pass
 
-    def get_commonlines(self):
+        if memory_factor < 0 or (memory_factor > 1 and memory_factor < 100):
+            logger.error('Subsampling factor must be between 0 and 1 or larger than 100.')
+
+        n_theta = self.n_theta // 2
+        n_img = self.n_img
+        rotations = self.rotations
+
+        pf = self.pf.copy()
+
+        # Estimate number of equations that will be used to calculate the shifts
+        n_equations = self._estimate_num_shift_equations(n_img, memory_factor)
+        # Allocate local variables for estimating 2D shifts based on the estimated number
+        # of equations. The shift equations are represented using a sparse matrix,
+        # since each row in the system contains four non-zeros (as it involves
+        # exactly four unknowns). The variables below are used to construct
+        # this sparse system. The k'th non-zero element of the equations matrix
+        # is stored at index (shift_i(k),shift_i(k)).
+        shift_i = np.zeros(5 * n_equations)
+        shift_j = np.zeros(5 * n_equations)
+        shift_eq = np.zeros(5 * n_equations)
+        shift_b = np.zeros(n_equations)
+
+        # Prepare the shift phases to try and generate filter for common-line detection
+        max_shift = self.max_shift
+        shift_step = self.shift_step
+        r_max = (pf.shape[0] - 1) // 2
+        _, shift_phases, h = self._generate_shift_phase_and_filter(r_max, max_shift, shift_step)
+
+        d_theta = np.pi / n_theta
+
+        # Generate two index lists for [i, j] pairs of images
```

Comment:
Done.

Reply:
But instead of generating all of them and then subsampling, can't this be done in one function? Seems like it's a bit wasteful to generate them all and throw most of them out.

Reply:
Done in the new version.